planner/core/runtime_filter.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/zap"
+)
+
+type RuntimeFilterType = variable.RuntimeFilterType
+type RuntimeFilterMode = variable.RuntimeFilterMode
+
+// RuntimeFilter structure is generated by the Runtime Filter Generator.
+// At present, it has a one-to-one correspondence with the equivalent expression in HashJoin.
+// For example:
+// Query: select * from t1, t2 where t1.k1=t2.k1
+// PhysicalPlanTree:
+//
+//	HashJoin_2(t1.k2=t2.k1)
+//	/           \
+//
+// TableScan_0(t1)     ExchangeNode_1
+// RuntimeFilter struct:
+//
+//	id: 0
+//	buildNode: HashJoin
+//	srcExprList: [t2.k1]
+//	targetExprList: [t1.k1]
+//	rfType:  IN
+//	rfMode: LOCAL
+//	buildNodeExplainId: HashJoin_2
+//	targetNodeExplainId: TableScan_0
+//
+// Although srcExprList and targetExprList attributes are lists, in fact, there will only be one value in them at present.
+// The reason why it is designed as a list is because:
+// Reserve an interface for subsequent multiple join equivalent expressions corresponding to one rf.
+// (Since IN supports multiple columns, this construction can be more efficient at the execution level)
+type RuntimeFilter struct {
+	// runtime filter id, unique in one query plan
+	id             int
+	buildNode      *PhysicalHashJoin
+	srcExprList    []*expression.Column
+	targetExprList []*expression.Column
+	rfType         RuntimeFilterType
+	// The following properties need to be set after assigning a scan node to RF
+	rfMode     RuntimeFilterMode
+	targetNode *PhysicalTableScan
+	// The explain id will be set when runtime filter clone()
+	// It is only used for runtime filter pb
```

Comment:
So, does this mean the `RuntimeFilter` must be cloned before being converted to pb?

Reply:
Currently yes, because as long as the task is sent to tiflash, ```clone``` will be called again.
Secondly, because the attribute explain id is actually used exclusively for the PB structure, it is not used before. 
So I delayed the construction of this property, so that the construction and use of the property should be as close as possible.